### DISCUSSION TITLE
Marking triple points of CSL grain boundaries

### DISCUSSION THREAD

**Discussion by ianejo** [mtex-toolbox/mtex#1474] [Ask Anything]
Hi all,

I tried to implement these 2 lines of code i got from MTEX documentation on an austenitic phase of a steel data set:

% logical list of CSL boundaries
isCSL3 = grains.boundary.isTwinning(CSL(3,ebsd('Fe fcc').CS),3*degree);

% logical list of triple points with at least 2 CSL boundaries
tPid = sum(isCSL3(grains('Fe fcc').triplePoints.boundaryId),2)>=2;

The first line was successful but the second line wasn't. The error returned was

"Index exceeds the number of array elements. Index must not exceed 18735"

How do i fix this?

Thanks.

**Comment by kilir** [comment]
Hi,
could you elaborate what exactly you want to do? Count how many twin boundaries meet at triple points?
Maybe you'll have a sample dataset and a little bit more of explanation what you like to achieve.

Edit: yes it makes sense, but potentially if you have several phases you need to use the `'id'` but not the index.

Cheers,
Rüdiger

**Comment by ianejo** [comment]
Hello,

So i have a steel data set comprising a ferrite(bcc) and austenite(fcc) phase.

I plotted the CSL (3) boundaries for each phase using the code on MTEX documentation successfully.

Next i wish to mark the triple points with at least 2 CSL boundaries and plot them using the code in the documentation:

% logical list of CSL boundaries
isCSL3 = grains(Fe fcc).boundary.isTwinning(CSL(3,ebsd.CS),3*degree);

% logical list of triple points with at least 2 CSL boundaries
tPid = sum(isCSL3(grains.triplePoints.boundaryId),2)>=2;

The first code worked but the second did not even after editing thus:

tPid = sum(isCSL3(grains('Fe fcc').triplePoints.boundaryId),2)

Instead i got the said error :

"Index exceeds the number of array elements. Index must not exceed 18735"

I admit i don't have much of coding experience, hence my plea for help.

Thanks.



________________________________


Hi,
could you elaborate what exactly you want to do? Count how many twin boundaries meet at triple points?

This line

sum(isCSL3(grains('Fe fcc').triplePoints.boundaryId),2)



does not really make sense because you want to sum up grain boundaries.

Maybe you'll have a sample dataset and a little bit more of explanation what you like to achieve.

Cheers,
Rüdiger

**Reply by kilir** [reply]
The easiest way to help would be, if you can provide a little dataset and the relevant lines till the point you get stuck.
Cheers,
Rüdiger

**Reply by ianejo** [reply]
Okay.

% grain reconstruction
[grains,ebsd('indexed').grainId] = calcGrains(ebsd('indexed'),'angle',10*degree);

% grain smoothing
grains = smooth(grains('Fe fcc'),5);

% plot the result
plot(grains,grains.meanOrientation)

gB = grains.boundary;
hold on
plot(gB,'lineWidth',1)
hold off


% Plotting the band contrast and overlaying it with orientation map
plot(ebsd('Fe bcc'),ebsd('Fe bcc').bc,"coordinates","on")
mtexColorMap gray

hold on
plot(grains,grains.meanOrientation,'FaceAlpha',0.1,'linewidth',1)
hold off


% Detecting and plotting CSL(3) boundaries for Fe-fcc phase
gB = grains.boundary('Fe fcc','Fe fcc');

% select CSL(3) grain boundaries
gB_3 = gB(angle(gB.misorientation,CSL(3,ebsd('Fe bcc').CS))3*degree));

% overlay CSL(3) grain boundaries with the existing plot
hold on
plot(gB_3,'lineColor','Red','linewidth',3,'DisplayName','CSL 3')
hold off

![CSL(3)](https://user-images.githubusercontent.com/108414623/177708437-7c3120b8-d8f1-4c06-8e6c-838d5deffbd4.png)



% Next, mark and plot triple points

THIS IS THE CODE IN MTEX DOCUMENTATION:

% logical list of CSL boundaries
isCSL3 = grains.boundary.isTwinning(CSL(3,ebsd.CS),3*degree);
% logical list of triple points with at least 2 CSL boundaries
tPid = sum(isCSL3(grains.triplePoints.boundaryId),2)>=2;
% plot these triple points
hold on
plot(grains.triplePoints(tPid),'color','red','linewidth',2,'MarkerSize',8)
hold off

SO I EDITED THE FIRST LINE OF CODE TO SUIT MY DATA THUS:

isCSL3 = grains.boundary.isTwinning(CSL(3,ebsd('Fe fcc').CS),3*degree);  % AND THIS LINE WAS EXECUTED SUCCESSFULLY!

THE ISSUE IS WITH THE SECOND LINE OF CODE:

tPid = sum(isCSL3(grains.triplePoints.boundaryId),2)>=2;     % I TRIED EDITING IN ORDER TO EXECUTE THIS LINE TO NO AVAIL!

So this is where i got stuck.

Thanks

**Reply by kilir** [reply]
Ok, that's all clear.
But still, what dataset is it?

**Reply by ianejo** [reply]
EBSD duplex steel dataset.

**Reply by kilir** [reply]
So which one exactly?
This is the list of ebsd data we have as example data.
```
>> mtexdata
available loading routines for mtex sample data
                              type                      files              
                         ______________    ________________________________

    aachen               {'EBSD'      }    {'85_829grad_07_09_06.txt'     }
    sharp                {'EBSD'      }    {'sharp.txt'                   }
    small                {'EBSD'      }    {'Forsterite.ctf'              }
    csl                  {'EBSD'      }    {'CSL.txt'                     }
    3d                   {'EBSD'      }    {'3dData/*.ANG'                }
    mylonite             {'EBSD'      }    {'P5629U1.txt'                 }
    epidote              {'EBSD'      }    {'data.ctf'                    }
    forsterite           {'EBSD'      }    {'Forsterite.ctf'              }
    olivine              {'EBSD'      }    {'olivineopticalmap.ang'       }
    twins                {'EBSD'      }    {'twins.ctf'                   }
    copper               {'EBSD'      }    {'copper.osc'                  }
    single               {'EBSD'      }    {'single_grain_aluminum.txt'   }
    alu                  {'EBSD'      }    {'polycrystalline_aluminum.txt'}
    titanium             {'EBSD'      }    {'titanium.txt'                }
    ferrite              {'EBSD'      }    {'ferrite.ang'                 }
    martensite           {'EBSD'      }    {'martensite.cpr'              }
    emsland              {'EBSD'      }    {'Emsland_plessite_500x_15.ctf'}
    alphaBetaTitanium    {'EBSD'      }    {'EDXLMDTi64.cpr'              }
```

If not, please provide a little example dataset.

Rüdiger

**Reply by ianejo** [reply]
The dataset is none of the Mtex example dataset.
I sent the dataset to you via email already. Thanks

**Reply by kilir** [reply]
Ok, your dataset is contains two phases, both called "Fe" which is probably not such a good idea. After renaming them to Fe1 and Fe2, the latter being the phase containing the CSL3 boundaries, you can search those triple points for example by something like that:

```
gB = grains.boundary;

% select CSL(3) grain boundaries
gB_3 = gB(gB.isTwinning(CSL(3,ebsd('Fe2').CS),3*degree))

plot(gB)
hold on
plot(gB_3,'linecolor','r','linewidth',2)

% triple point candidates of two cls3 boundaries 
tPC = gB_3.triplePoints('Fe2','Fe2','Fe2')


id = sum(ismember(tPC.nextVertexId,unique(gB_3.F(:))),2) > 1 % works unless two adjacent tPs share 1 segment

plot(tPC(id),'color','g','linewidth',2,'MarkerSize',8)
hold off
```
Cheers,
Rüdiger

**Reply by ianejo** [reply]
The above code did the trick.

Much gratitude Rudiger!
